### PR TITLE
(all): convert deck to hybrid ng1/ng2 app

### PR DESCRIPTION
**Pull request by icfantv**
* switch to ng2 bootstrapping of application
* upgrade to webpack 2
* upgrade to typescript 2.1.x
* switch over to awesome-typescript-loader
* update codelyzer and tslint to latest and update rules
* sanitize application to fix all current linting (JS/TS) issues.

### CODE REVIEW

**Comment by icfantv** [comment]
@anotherchrisberry @zanthrash @jrsquared @danielpeach PTAL

**NB** run `npm run clean`, delete your `.happypack` directory (not strictly necessary), and run `npm install` before kicking the tires on this.

File: app/scripts/modules/core/pipeline/config/validation/pipelineConfig.validator.spec.ts [hunk 1]
```diff
@@ -88,7 +89,7 @@ describe('pipelineConfigValidator', () => {
   }));
 
   beforeEach(mock.inject((_pipelineConfigValidator_: PipelineConfigValidator,
-                          _pipelineConfig_: PipelineConfigService,
+                          _pipelineConfig_: any,
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/pipeline/config/validation/pipelineConfig.validator.spec.ts]: Is `PipelineConfigService` not valid here?

Comment (icfantv) [inline app/scripts/modules/core/pipeline/config/validation/pipelineConfig.validator.spec.ts]: So this is kind of badass/PITA.  The compiler is actually checking that the `spyOn` calls refer to methods that actually exist on the object which the dev claims they're being spied upon.  In this case, the `spyOn` calls are actually spying on the `.provider('pipelineConfig')` which is not TS and `angular.auto.IProvideService` is not templatized so it doesn't know the type.  Either way, however, I think PipelineConfigService wasn't correct, originally.

**Comment by anotherchrisberry** [comment]
Page title is not updating (just shows as "Spinnaker")

**Comment by anotherchrisberry** [comment]
Is anything actually converted to ng2 here? I see a dummy service, but no controllers, components, directives, etc.

**Comment by icfantv** [comment]
No and I can explain in more detail tomorrow.  My intention is try and get the authentication service working again but wanted the framework in place.  I have concerns about other service conversions.

**Comment by icfantv** [comment]
@anotherchrisberry @jrsquared @zanthrash @danielpeach updated to convert `AuthenticationService` to ng2 and wire into the application/tests.  Take a look at `authentication.interceptor.spec.ts` to see the solution @robwormald recommended we use instead of trying to shoehorn in an mocked application given our current dependency graph and the errors we saw on the upgrade route.

This should provide us a way to move forward for existing code whereas for new ng2 code we should be able to use `dummyNg1.service.spec.ts` as the pattern moving forward.  I want to take another crack at upgrading `RedirectService` to capture the issues and possibly loop in @petebacondarwin and @gkalpak to see if it's on our end or theirs.

File: app/scripts/modules/core/config/settings.js [hunk 2]
```diff
@@ -2,5 +2,6 @@
 
 let angular = require('angular');
 
-module.exports = angular.module('spinnaker.core.config.settings', [])
+export const SETTINGS = 'spinnaker.core.config.settings';
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/config/settings.js]: Doesn't look `SETTINGS` is being imported anywhere. In the interest of keeping the PR succinct, maybe revert this particular change?

File: app/scripts/modules/core/upgrade/dummyNg1.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        expect(dummyNg2Service.getMessage()).toBe('Dummy NG2 Service');
+
+        const dummyNg1Service: DummyNg1Service = injector.get('dummyNg1Service') as DummyNg1Service;
+        expect(dummyNg1Service).toBeDefined();
+        expect(dummyNg1Service.getMessage()).toBe('Dummy NG1 Service');
+        expect(dummyNg1Service.getInjectedMessage()).toBe('Dummy NG2 Service');
+        done();
+      });
+  });
+});
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/upgrade/dummyNg1.service.spec.ts]: With authentication converted/downgraded, does it make sense to remove the dummy services, or are they still providing value on how to wire up ng2 components?

**Comment by anotherchrisberry** [comment]
Page title is still not being updated (it's set via `$rootScope`, which doesn't really work now). You can probably get around it by tweaking the `pageTitleService` to set `$document[0].title` wherever it's setting `$rootScope.pageTitle`.

**Comment by anotherchrisberry** [comment]
Other than the page title piece, I'm able to run the app and tests with no noticeable difference in functionality or performance. `npm start` is a minute slower, but that's expected, and we can probably work on that later.

I would still really like to see a component converted to ng2 and see how that gets wired in/used before merging this.

File: app/scripts/modules/core/config/settings.js (hunk 2 above)
Comment (icfantv) [inline app/scripts/modules/core/config/settings.js]: Yup, extraneous artifact.  Will remove.

File: app/scripts/modules/core/upgrade/dummyNg1.service.spec.ts (hunk 3 above)
Comment (icfantv) [inline app/scripts/modules/core/upgrade/dummyNg1.service.spec.ts]: Yea, I was thinking the latter, if only temporarily.  It's definitely cruft that can be removed, but I didn't want to lose it.

**Comment by icfantv** [comment]
I missed the title bit, sorry about that.  I'll take a look at it.  RE downgraded component, let me find a small one.

**Comment by icfantv** [comment]
@anotherchrisberry title updating is fixed and i've converted a component to NG2 and downgraded it/wired it in.  This exposed some pain points we will have moving forward surrounding AoT but _might_ be mitigable via some sort of utility functions.

File: app/scripts/modules/amazon/subnet/subnetSelectField.component.ts [hunk 4]
```diff
@@ -86,3 +86,5 @@ export const SUBNET_SELECT_FIELD_COMPONENT = 'spinnaker.amazon.subnet.subnetSele
 module(SUBNET_SELECT_FIELD_COMPONENT, [
   require('core/config/settings'),
 ]).component('subnetSelectField', new SubnetSelectFieldComponent());
+
+
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/subnet/subnetSelectField.component.ts]: extraneous whitespace

**Comment by anotherchrisberry** [comment]
Pulled this down and ran it locally and everything seems to work.

I'm good with merging this - @danielpeach or @jrsquared or @zanthrash any concerns?

**Comment by jrsquared** [comment]
Just pulled it down too.  I say now that we have working examples of how to convert, let's get this in and fix any minor issues as they come. 👍

**Comment by icfantv** [comment]
removed extraneous whitespace.

**Pull request merged**